File: mac/Sources/Zap/Enrollment.swift
import Foundation

/// A connection request, saved with the identity before it is sent. The server answers a repeat of the
/// exact request with the original credentials, so an interrupted attempt can be resent safely.
struct PendingEnrollment: Codable {
    var url: String
    var path: String
    var credential: String
    var body: Data
    var keys: [String: Data]
    var keyId: String

    init(url: String, path: String, credential: String = "", registration: [String: Any], keys: [String: Data], keyId: String) throws {
        self.url = url
        self.path = path
        self.credential = credential
        self.keys = keys
        self.keyId = keyId
        var registration = registration
        registration["enrollmentId"] = UUID().uuidString
        registration["deviceToken"] = VaultCrypto.randomKey().map { String(format: "%02x", $0) }.joined()
        body = try JSONSerialization.data(withJSONObject: registration)
    }

    func completed(_ identity: Identity, response: Data) throws -> Identity {
        let credentials = try JSONDecoder().decode(EnrollmentCredentials.self, from: response)
        let request = try JSONSerialization.jsonObject(with: body) as? [String: Any]
        guard credentials.deviceId == request?["enrollmentId"] as? String,
              credentials.token == request?["deviceToken"] as? String,
              credentials.keyId == keyId
        else { throw ZapError("The server returned a different enrollment. Update the server and retry.") }

        var next = identity
        next.url = url
        next.token = credentials.token
        next.deviceId = credentials.deviceId
        next.keys = keys
        next.keyId = keyId
        next.enrollment = nil
        return next
    }
}

private struct EnrollmentCredentials: Decodable {
    var deviceId: String
    var token: String
    var keyId: String
}

File: android/app/src/main/java/dev/midplane/zap/Enrollment.kt
package dev.midplane.zap

import org.json.JSONObject
import java.util.Base64
import java.util.UUID

const val PAIRING_CODE_PREFIX = "zap://pair#"

fun decodePairingCode(code: String): JSONObject =
    JSONObject(String(Base64.getUrlDecoder().decode(code.substringAfter('#'))))

fun encodePairingCode(pairing: JSONObject): String =
    PAIRING_CODE_PREFIX + Base64.getUrlEncoder().withoutPadding().encodeToString(pairing.toString().toByteArray())

/**
 * A connection request, stored inside the encrypted identity before it is sent. The server answers a repeat
 * of the exact request with the original credentials, so an interrupted attempt can be resent safely.
 */
class PendingEnrollment(val record: JSONObject) {
    val url: String get() = record.getString("url")
    val body: JSONObject get() = record.getJSONObject("body")

    fun completed(response: JSONObject): Array<Pair<String, Any>> {
        val matches = response.getString("deviceId") == body.getString("enrollmentId") &&
            response.getString("token") == body.getString("deviceToken") &&
            response.getString("keyId") == body.getString("keyId")
        require(matches) { "The server returned a different enrollment. Update the server and retry." }
        return arrayOf(
            "url" to url,
            "token" to response.getString("token"),
            "deviceId" to response.getString("deviceId"),
            "keyId" to body.getString("keyId"),
            "keys" to record.getJSONObject("keys"),
            "enrollment" to ""
        )
    }

    companion object {
        fun create(url: String, registration: JSONObject, keys: JSONObject): PendingEnrollment {
            val body = JSONObject(registration.toString())
                .put("enrollmentId", UUID.randomUUID().toString())
                .put("deviceToken", Crypto.randomKey().joinToString("") { "%02x".format(it) })
            return PendingEnrollment(JSONObject().put("url", url).put("body", body).put("keys", JSONObject(keys.toString())))
        }
    }
}
